include/hip/hcc_detail/program_state.hpp

```diff
@@ -558,92 +561,156 @@ const std::unordered_map<
 }
 
 inline
-std::size_t parse_args(
-    const std::string& metadata,
-    std::size_t f,
-    std::size_t l,
-    std::vector<std::pair<std::size_t, std::size_t>>& size_align) {
-    if (f == l) return f;
-    if (!size_align.empty()) return l;
-
-    do {
-        static constexpr size_t size_sz{5};
-        f = metadata.find("Size:", f) + size_sz;
-
-        if (l <= f) return f;
-
-        auto size = std::strtoul(&metadata[f], nullptr, 10);
-
-        static constexpr size_t align_sz{6};
-        f = metadata.find("Align:", f) + align_sz;
-
-        char* l{};
-        auto align = std::strtoul(&metadata[f], &l, 10);
-
-        f += (l - &metadata[f]) + 1;
+void checkError(
+    amd_comgr_status_t status,
+    char const *str) {
+    if (status != AMD_COMGR_STATUS_SUCCESS) {
+        const char *status_str;
+        status = amd_comgr_status_string(status, &status_str);
+        std::cerr << "FAILED: " << str << "\n  REASON: " <<  status_str << std::endl;
+        hip_throw(std::runtime_error{"Metadata parsing failed."});
+    }
+}
 
-        size_align.emplace_back(size, align);
-    } while (true);
+inline
+std::string lookup_keyword_value(
+    amd_comgr_metadata_node_t& in_node,
+    std::string keyword) {
+    amd_comgr_status_t status;
+    size_t value_size;
+    amd_comgr_metadata_node_t value_node;
+
+    status = amd_comgr_metadata_lookup(in_node, keyword.c_str(), &value_node);
+    checkError(status, "amd_comgr_metadata_lookup");
+    status = amd_comgr_get_metadata_string(value_node, &value_size, NULL);
+    checkError(status, "amd_comgr_get_metadata_string");
+    // Since value_size returns size with null terminator, we don't include for C++ string size
+    value_size--;
+    std::string value(value_size, '\0');
+    status = amd_comgr_get_metadata_string(value_node, &value_size, &value[0]);
+    checkError(status, "amd_comgr_get_metadata_string");
+
+    status = amd_comgr_destroy_metadata(value_node);
+    checkError(status, "amd_comgr_destroy_metadata");
+    return value;
 }
 
 inline
-void read_kernarg_metadata(
-    ELFIO::elfio& reader,
+void process_kernarg_metadata(
+    amd_comgr_metadata_node_t blob_meta,
     std::unordered_map<
         std::string,
         std::vector<std::pair<std::size_t, std::size_t>>>& kernargs) {
-    // TODO: this is inefficient.
-    auto it = find_section_if(reader, [](const ELFIO::section* x) {
-        return x->get_type() == SHT_NOTE;
-    });
+    amd_comgr_status_t status;
+    amd_comgr_metadata_kind_t mkindLookup;
+    amd_comgr_metadata_node_t kernelList, kernelMap;
+    amd_comgr_metadata_node_t kernArgList, kernArgMap;
+    std::string kernel_name;
+    size_t num_kernels = 0;
+    size_t num_kern_args = 0;
+
+    // Kernels is a list of MAPS!!
+    status = amd_comgr_metadata_lookup(blob_meta, "Kernels", &kernelList);
+    if(status != AMD_COMGR_STATUS_SUCCESS)
+        return;
```

Comment:
Why `return` here, rather than throw? The "Kernels" key is required (see https://llvm.org/docs/AMDGPUUsage.html#amdgpu-amdhsa-code-object-metadata-map-table-v2).

Reply:
There seem to be some HIP memset kernels which don't have a "Kernels" key. So if I throw here, all those kernels that are missing this field will fail to run. @mangupta, I believe that makes them invalid kernels, should we create a different PR to fix those kernels?

Reply:
I would imagine that is a bug somewhere else, and allowing it here will probably just be papering over that? I'm not sure what other consumers do currently.

Reply:
I've added a FIXME for the HIP team to fix those kernels in the future. It should be fixed in a separate PR since its out of this PR's context. Thanks!